Move SANS gui property handling to a separate class

TTitcombe opened the pull request:
**Description of work.**
sans_data_processor_gui and add_runs_page, two gui pages within the ISIS SANS interface, use qt properties to set default properties on loading of the gui. The functionality to achieve this has been moved to a separate class to clean up the code, and to make it easier for any other gui page to set and grab properties in the future

This PR also moves a config observer from the add_runs_presenter to run_tab_presenter, as we no longer need it in the add_runs_presenter

**To test:**
1. Interfaces -> SANS -> SANS ISIS
2. On the main gui, select an output mode (one of Memory, File, Both) and toggle Save Can
3. In the add runs page, select an output folder
4. Close and re-open the gui. The settings you just changed should have been remembered

Fixes partially #24529

*This does not require release notes* because **internal code change**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

TTitcombe committed:
separate default properties from SANSDataProcessorGUI

Move OutputPropertyObserver from add_runs_pres to run_tab_pres

TTitcombe committed:
Add methods to update properties in SANSGuiPropertiesHandler

Use SANSGuiPropertiesHandler in sans_data_processor_gui

TTitcombe committed:
Remove unused has_any_event_data import in add_runs_presenter

TTitcombe committed:
Fix flake8 unused import and indentation errors

TTitcombe committed:
Add unit tests to check add runs presenter loads and sets output dir

TTitcombe committed:
Fix run_tab_presenter_test by mocking out line edit

Stop passing logger into propertyHandler

TTitcombe committed:
Add unit test for SANSGuiPropertiesHandler

mantid-builder commented:
@TTitcombe it may be worth bringing this PR to attention for review.
<!-- {"problem_type": "unreviewed"} -->

TTitcombe committed:
Merge branch 'master' into 24529_separate_default_gui_properties

hsautessella commented on scripts/SANS/sans/gui_logic/gui_common.py:
```diff
@@ -4,11 +4,14 @@
 import os
 
+from qtpy.QtCore import (QSettings)  # noqa
+from qtpy.QtWidgets import (QFileDialog)  # noqa
```
Do we need the brackets and `noqa` on these imports?

TTitcombe replied:
I don't think so, I don't know why I added them

hsautessella commented on scripts/SANS/sans/gui_logic/presenter/add_runs_presenter.py:
```diff
@@ -134,13 +123,15 @@ def _handle_selection_changed(self, run_selection):
         directory = self._view.display_save_directory_box("Save sum runs", self.save_directory)
-        self._update_output_directory(directory)
-        self._view.set_out_file_directory(self.save_directory)
+        directory = os.path.join(directory, '')  # Add an OS specific trailing slash if it doesn't already exist
```
As a more explicit alternative you could use `directory + os.sep`, but the `.join` with the comment is just as good

TTitcombe replied:
using join will only add a slash if it doesn't end in one. I had noticed that sometimes when you selected an output directory it would end in a slash and other times it wouldn't. 

I didn't know about os.sep, though, so I'll keep that in mind for cases where we always need to add one

hsautessella commented on scripts/SANS/sans/gui_logic/presenter/run_tab_presenter.py:
```diff
@@ -302,6 +309,11 @@ def set_view(self, view):
+                                                                                        str)},
+                                                                         line_edits={"user_file":
+                                                                                     self._view.user_file_line_edit})
+
```
You could move these arguments to a new line to avoid the PEP8 line length warning

hsautessella commented on scripts/test/SANS/gui_logic/add_runs_presenter_test.py:
```diff
@@ -431,4 +432,38 @@ def test_that_make_filename_returns_fullname_if_one_is_given(self):
+        self.presenter.set_output_directory = mock.Mock()
+
+        self.presenter._handle_output_directory_changed()
+        self.assertEqual(self.presenter.gui_properties_handler.update_default.call_count, 1,
```
You could use mock's `assert_called_once_with` here, which also checks the arguments the function is called with. E.g. 

```python
self.presenter.gui_properties_handler.update_default.assert_called_once_with('add_runs_output_directory', new_dir_name)

TTitcombe replied:
That's a far better solution

TTitcombe committed:
Fix reviewer comments

Use assert_called_once_with in unit tests

Fix flake8 line length warning

Remove unnecessary noqa comments on imports
